daemon/daemon/swap_in.go

```diff
@@ -0,0 +1,165 @@
+			logger.Debug("failed. The funds have been refunded")
+		case models.OutcomeExpired:
+			outcome := models.OutcomeExpired
+			currentSwap.Outcome = &outcome
+			logger.Debug("failed. The contract has expired, waiting to be refunded")
+		default:
+			outcome := models.OutcomeSuccess
+			currentSwap.Outcome = &outcome
+			// FAILED doesn't exist in the 40swap backend so we don't need to check it
+			logger.Debug("success. The funds have been claimed")
+		}
+	case models.StatusContractRefundedUnconfirmed:
+		log.Debug("the refund has been sent, waiting for on-chain confirmation")
+	case models.StatusContractExpired:
+		if currentSwap.RefundRequestedAt.IsZero() { // check refund was requested
+			currentSwap.RefundRequestedAt = m.now()
+			log.Info("on-chain contract expired. initiating a refund")
+			txId, err := m.InitiateRefund(ctx, currentSwap)
+			if err != nil {
+				return fmt.Errorf("failed to initiate refund: %w", err)
+			}
+			currentSwap.RefundTxID = txId
+		} else {
+			log.Debug("on-chain contract expired. Refund is in-progress")
+		}
+	}
+
+	if changed {
+		currentSwap.Status = newStatus
+		err := m.repository.SaveSwapIn(&currentSwap)
+		if err != nil {
+			return fmt.Errorf("failed to save swap in: %w", err)
+		}
+	}
+
+	logger.Debug("swap in processed")
+
+	return nil
+}
+
+func (m *SwapMonitor) InitiateRefund(ctx context.Context, swap models.SwapIn) (string, error) {
+	logger := log.WithFields(log.Fields{
+		"swap_id": swap.SwapID,
+	})
+
+	logger.Infof("Claiming swap in refund: %s", swap.SwapID)
+	res, err := m.swapClient.GetRefundPSBT(ctx, swap.SwapID, swap.RefundAddress)
+	if err != nil {
+		return "", fmt.Errorf("failed to get refund psbt: %w", err)
+	}
+
+	pkt, err := bitcoin.Base64ToPsbt(res.PSBT)
+	if err != nil {
+		return "", fmt.Errorf("failed to decode psbt: %w", err)
+	}
+
+	// check if the refund address returned in the psbt is our own
+	if !isValidRefundTx(pkt, m.network, swap.RefundAddress) {
+		return "", fmt.Errorf("invalid refund tx")
+	}
+
+	privateKey, err := bitcoin.DeserializePrivateKey(swap.RefundPrivatekey)
+	if err != nil {
+		return "", fmt.Errorf("failed to decode refund private key: %w", err)
+	}
+
+	// Process the PSBT
+	tx, err := bitcoin.SignFinishExtractPSBT(logger, pkt, privateKey, &lntypes.Preimage{}, 0)
+	if err != nil {
+		return "", fmt.Errorf("failed to sign PSBT: %w", err)
+	}
+
+	if fee, err := pkt.GetTxFee(); err != nil || fee > 1000 {
+		return "", fmt.Errorf(`fee rate too high ${psbt.getFeeRate()}`)
+	}
+
+	serializedTx, err := bitcoin.SerializeTx(tx)
+	if err != nil {
+		return "", fmt.Errorf("failed to serialize transaction: %w", err)
+	}
+
+	// Send transaction back to the swap client
+	logger.Debug("Sending transaction back to swap client")
+	err = m.swapClient.PostRefund(ctx, swap.SwapID, serializedTx)
+	if err != nil {
+		return "", err
+	}
+
+	return tx.TxID(), nil
+}
+
+func isValidRefundTx(psbt *psbt.Packet, network lightning.Network, address string) bool {
```

Comment:
Could you please move this to bitcoin pkg so I can use it on swap outs too?

Reply:
done